---
title: "Structural analysis"
# knit: (function(input_file, encoding) {
#   rmarkdown::render(input_file,
date: "`r format(Sys.time(), '%d-%b-%Y')`"
author: "Giacomo Mutti"
output:
    rmdformats::html_clean:
        number_sections: false
        code_folding: hide
        self_contained: true
        fig_caption: true
        # thumbnails: false
        # lightbox: false
        gallery: true
        toc_depth: 3
        highlight: kate
---

```{r setup, include=FALSE, dev="CairoPNG"}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE, cache.lazy = FALSE,
                      crop = TRUE, dev="CairoPNG")
knitr::knit_hooks$set(crop = knitr::hook_pdfcrop)
knitr::opts_knit$set(root.dir = "../../")
```

```{r libraries, message=FALSE, warning=FALSE, cache=FALSE}
library(tidyverse)
library(patchwork)
library(ggtree)
library(cowplot)
library(glue)
source("workflow/scripts/functions.R")
theme_set(theme_classic())

models <- names(palettes_model)

params <- yaml::read_yaml("config/params.yaml")
params_dataset <- yaml::read_yaml("config/Hsapopi_test.yaml")

outdir <- glue(params$outdir, params_dataset$dataset)
```

```{r read_stuff, cache=FALSE}
seeds <- readLines(glue(outdir, "/ids/UP000005640_common.ids"))

tax <- read_delim(glue(outdir, "/meta/", params_dataset$dataset, "_uniprot_genomes.tsv"), 
                  show_col_types=F)
taxmap <- read_delim(glue(outdir, "/db/taxidmap"), 
                     delim = "\t", col_names = c("label", "Tax_ID"),
                     show_col_types = FALSE) %>% 
  left_join(tax, by = "Tax_ID") %>% 
  dplyr::select(label, Proteome_ID, mnemo)

taxidmap <- read_delim(glue(outdir,"/db/taxidmap"), 
                       col_names = c("target", "Tax_ID"), show_col_types=F)

sptree <- read.tree(params_dataset$species_tree)
translate <- deframe(distinct(dplyr::select(taxmap, Proteome_ID, mnemo)))
# sptree$tip.label <- names(translate)[match(translate, sptree$tip.label)]

species_dist_df <- tibble()
for (sp in sptree$tip.label){
  if (sp=="Hsap") {next}
  dist <- castor::get_all_distances_to_tip(sptree, "Hsap")[MRCA(sptree, "Hsap", sp)]
  mrca <- c(sptree$tip.label, sptree$node.label)[MRCA(sptree, "Hsap", sp)]
  species_dist_df <- bind_rows(species_dist_df, tibble(mnemo=sp, value=dist, mrca=mrca))
}

# read input table
# lvls_tax <- rev(c(unique(tax$Clade), "archaea", "bacteria"))

table_columns <- c('Proteome_ID', 'Tax_ID', 'count1', 'count2', 'count3', 
                   'genome', 'source', 'species', 'mnemo')

# table <- read_delim(params_dataset$input_table_file, show_col_types=F, 
#                     delim = "\t", col_names = table_columns) %>% 
#   left_join(tax, by = "Tax_ID") %>% 
#   mutate(Clade = ifelse(is.na(Clade), mnemo, Clade),
#          Clade = factor(Clade, levels = lvls_tax))

lddt_df <- read_delim(paste0(params$data_dir, "structures/", 
                             tax$Proteome_ID, "/", tax$Proteome_ID, "_mean_plddt.tsv"),
           col_names = c("target", "mean_lddt"), show_col_types = FALSE)

aln_seeds <- readLines(glue(outdir,"/ids/", params_dataset$seed, "_aln.ids"))
```

# Homology

```{r read_self_blast}
blast_self <- read_delim(glue(outdir, "/homology/allvall/",
                              params_dataset$seed, "_", params_dataset$seed, "_blast.tsv"),
                         show_col_types=FALSE,
                         col_names = blast_columns) %>% 
  # filter(query %in% seeds) %>%
  filter(query==target) %>% 
  mutate(blast_selfbit=bitscore) %>% 
  select(query, blast_selfbit)

fs_self <- read_delim(glue(outdir, "/homology/allvall/",
                              params_dataset$seed, "_", params_dataset$seed, "_fs.tsv"), 
                      show_col_types = FALSE,
                         col_names = blast_columns) %>% 
  # filter(query %in% seeds) %>%
  filter(query==target) %>% 
  mutate(fs_selfbit=bitscore) %>% 
  select(query, fs_selfbit)

self <- full_join(blast_self, fs_self, by="query")
```


```{r read_fs}
fs <- read_delim(glue(outdir, "/homology/", params_dataset$seed, "_fs.tsv"), 
                 col_names = fs_columns, show_col_types = FALSE) %>% 
  # filter(query %in% seeds) %>%
  mutate(evalue = ifelse(evalue==0, 1e-180, evalue)) %>%
  left_join(taxidmap, by = "target")

fs_brhs <- read_delim(glue(outdir, "/homology/", params_dataset$seed, "_fs_brh.tsv"),
                     col_names = c("query", "target"), show_col_types = FALSE) %>% 
  # filter(query %in% seeds) %>%
  mutate(method="fs_brh") 
```


```{r read_blast}
blast <- read_delim(glue(outdir, "/homology/", params_dataset$seed, "_blast.tsv"), 
                    col_names = blast_columns, show_col_types = FALSE) %>% 
  # filter(query %in% seeds) %>%
  mutate(evalue = ifelse(evalue==0, 1e-180, evalue))

blast_brhs <- read_delim(glue(outdir, "/homology/", params_dataset$seed, "_blast_brh.tsv"),
                     col_names = c("query", "target"), show_col_types = FALSE) %>% 
  # filter(query %in% seeds) %>%
  mutate(method="blast_brh")
```

```{r merge_homo}
df <- full_join(blast, fs,
                by = c("query", "target"), suffix=c("_Blast", "_Foldseek")) %>%
    filter(query!=target) %>% 
  left_join(self, by="query") %>%
  mutate(pident_Foldseek = 100*pident_Foldseek, 
         qcov_Foldseek = 100*qcov_Foldseek,
         singleton = case_when(is.na(evalue_Foldseek) ~ "only Blast",
                               is.na(evalue_Blast) ~ "only Foldseek", 
                               .default = "Common")) 

brhs <- rbind(blast_brhs, fs_brhs) %>% 
    group_by(query, target, method) %>%
    mutate(n=1) %>%
    pivot_wider(names_from = method, values_fill = 0, values_from = n) %>%
    mutate(common = ifelse(fs_brh+blast_brh==2, 1, 0))


blast_filt <- filter(blast, 
                         # query %in% paste0("AF-", aln_seeds, "-F1"),
                         length/qlen*100>params$coverage, 
                         length/slen*100>params$coverage,
                         evalue<as.numeric(params$eval_both)) %>% 
  group_by(query) %>% 
  slice_head(n = params$max_seqs) %>% 
  ungroup()

fs_filt <- filter(fs, 
                      # query %in% paste0("AF-", aln_seeds, "-F1"),
                      qcov*100>params$coverage, 
                      tcov*100>params$coverage, 
                      evalue<as.numeric(params$eval_both)) %>% 
  group_by(query) %>% 
  slice_head(n = params$max_seqs) %>% 
  ungroup()

df_filt <- full_join(blast_filt, fs_filt,
                by = c("query", "target"), suffix=c("_Blast", "_Foldseek")) %>%
    filter(query!=target) %>% 
  left_join(self, by="query") %>%
  mutate(pident_Foldseek = 100*pident_Foldseek, 
         qcov_Foldseek = 100*qcov_Foldseek,
         singleton = case_when(is.na(evalue_Foldseek) ~ "only Blast",
                               is.na(evalue_Blast) ~ "only Foldseek", 
                               .default = "Common"))
# grouped_brhs <- brhs %>%
#   group_by(query) %>%
#   summarise(fs_brh = sum(fs_brh), blast_brh=sum(blast_brh), common_brh=sum(common))
blast_filtered <- filter(blast, 
                         query %in% paste0("AF-", aln_seeds, "-F1"),
                         length/qlen*100>params$coverage, 
                         length/slen*100>params$coverage,
                         evalue<as.numeric(params$eval_both)) %>% 
  group_by(query) %>% 
  slice_head(n = params$max_seqs) %>% 
  ungroup()

fs_filtered <- filter(fs, 
                      query %in% paste0("AF-", aln_seeds, "-F1"),
                      qcov*100>params$coverage, 
                      tcov*100>params$coverage, 
                      evalue<as.numeric(params$eval_both)) %>% 
  group_by(query) %>% 
  slice_head(n = params$max_seqs) %>% 
  ungroup()

df_filtered <- full_join(blast_filtered, fs_filtered,
                by = c("query", "target"), suffix=c("_Blast", "_Foldseek")) %>%
    filter(query!=target) %>% 
  left_join(self, by="query") %>%
  mutate(pident_Foldseek = 100*pident_Foldseek, 
         qcov_Foldseek = 100*qcov_Foldseek,
         singleton = case_when(is.na(evalue_Foldseek) ~ "only Blast",
                               is.na(evalue_Blast) ~ "only Foldseek", 
                               .default = "Common"))
```


```{r cath}
CATH_df <- read_delim(list.files(glue(params$data_dir, "cath"), full.names = TRUE),
                      delim = "\t", id = "proteome", show_col_types = FALSE) %>%
  # filter(!is.na(Gene3D)) %>% 
  mutate(proteome = gsub("_cath.tsv", "", basename(proteome)),
         n_3d_domains = str_count(Gene3D, ";"),
         n_Pfam_domains = str_count(Pfam, ";")) %>%
  # filter(n_domains==1) %>% 
  mutate(target=paste0("AF-", Entry, "-F1")) %>% 
    select(-proteome, Entry)

nms <- read_delim("resources/cath-names.txt", 
                  comment = "#", delim = "    ",
                  col_names = c("Gene3D", "abbr", "name"), 
                  show_col_types = FALSE) %>% 
  select(-abbr) %>% 
  mutate(name=gsub("^:", "", name))

nms_pfam <- read_delim("resources/Pfam-A.clans.tsv", 
           col_names = c("Pfam", "clan", "boh", "abbr", "name"), 
           show_col_types = FALSE) %>% 
  select(-name, -clan, -boh)

long_CATH_df <- CATH_df %>% 
  # filter(target %in% unique(blast$query)) %>% 
  mutate(Gene3D=strsplit(Gene3D, ";")) %>% 
  unnest(Gene3D) %>% 
  separate(col = Gene3D, into = c("C", "A", "T", "H"), sep = "\\.", remove = FALSE) %>% 
  mutate(CA = paste(C,A,sep = "."), CAT=paste(C,A,T,sep = ".")) %>% 
  left_join(nms, by=c("C"="Gene3D")) %>%  
  left_join(nms, by=c("CA"="Gene3D"), suffix = c("_C", "_CA")) %>% 
  left_join(nms, by=c("CAT"="Gene3D"), suffix = c("_CA", "_CAT")) %>% 
  left_join(nms, by=c("Gene3D"), suffix = c("_CAT", "_CATH"))

long_CATH_nms <- long_CATH_df %>%
    select(-c(Entry, Pfam, n_Pfam_domains, target, n_3d_domains)) %>% 
    distinct()
```

## Fig1

```{r fig1}
fig1a <- df %>%
  group_by(singleton) %>%
  summarise(n = n()) %>%
  ungroup() %>%
  mutate(prop=n/sum(n), Filtering="Before") %>%
  bind_rows(df_filt %>%
            group_by(singleton) %>%
            summarise(n = n()) %>%
            ungroup() %>%
            mutate(prop=n/sum(n), Filtering="After")) %>%
  mutate(Filtering=factor(Filtering, labels = c("After", "Before"), ordered = T),
         singleton=case_when(singleton=="only Blast"~"Only Bp", 
                             singleton=="only Foldseek" ~ "Only Fs",
                             .default = singleton)) %>%
  ggplot(aes(singleton, n, fill=singleton)) +
  geom_bar(aes(alpha=Filtering), stat = "identity", position = position_dodge2(reverse=TRUE)) +
  geom_text(aes(group=Filtering, label=paste0(round(prop, 2)*100, "%")),
            position = position_dodge2(reverse = TRUE, width = 0.9),
            vjust=1.3, color="white", size=3) +
  scale_fill_manual(values = palette_singleton_s, guide="none") +
  scale_alpha_manual(values = c(1,.6)) +
  scale_y_continuous(expand = c(0,0), labels = scales::label_number(scale = 1e-6)) +
  labs(fill="Class", x="Class", y="# aligned protein pairs (Mln)") +
  theme(legend.position = "bottom", legend.box="vertical")

fig1b <- lddt_df %>% 
  ggplot(aes(mean_lddt)) + 
  geom_density(fill="#BF4375", size=1) +
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  labs(x="Mean pLDDT", y="") +
  geom_vline(xintercept = 40, linetype=2) +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank(),
        axis.line.y = element_blank())

ggsave("draft/figures/fig1_lddt.pdf", fig1b, width = 3, height = 4)
ggsave("draft/figures/fig1_cnts.pdf", fig1a, width = 4, height = 2)
```

## Fig2

```{r fig2}
eval_p <- ggplot(df, aes(-log10(evalue_Blast), -log10(evalue_Foldseek))) + 
  stat_bin_2d(breaks=seq(0,20,0.4)) + 
  scale_fill_gradientn(colors=wespal) +
  ggpubr::stat_cor(label.x.npc = .025) +
  # geom_hline(yintercept = -log10(evals), linetype=2) + 
  # geom_vline(xintercept = -log10(evals), linetype=2) + 
  coord_fixed(expand = 0) +
  labs(x="-log10(Evalue BlastP)",y="-log10(Evalue Foldseek)") +
  scale_x_continuous(limits = c(0, 20)) +
  scale_y_continuous(limits = c(0, 20)) +
  theme(legend.position = "none", panel.grid = element_blank()) 
eval_p <- add_marginal(mutate(df, 
                              evalue_Blast=-log10(evalue_Blast), 
                              evalue_Foldseek=-log10(evalue_Foldseek)), 
                       eval_p, "evalue_Blast", "evalue_Foldseek", "singleton", palette_singleton_p)

pident_p <- ggplot(df, aes(pident_Blast, pident_Foldseek)) + 
  stat_bin_2d(breaks=seq(0,100,2)) + 
  scale_fill_gradientn(colors=wespal) +
  ggpubr::stat_cor(label.x.npc = .025) +
  # scale_fill_viridis_c(option = "C") +
  coord_fixed(expand = 0) +
  scale_x_continuous(limits = c(0, 100)) +
  scale_y_continuous(limits = c(0, 100)) +
  labs(x="% Identity BlastP",y="% Identity Foldseek") +
  theme(legend.position = "none", panel.grid = element_blank())
pident_p <- add_marginal(df, pident_p, "pident_Blast", "pident_Foldseek", "singleton", palette_singleton_p)


qcov_p <- ggplot(df, aes(qcov_Blast, qcov_Foldseek)) + 
  stat_bin_2d(breaks=seq(0,100,2)) + 
  ggpubr::stat_cor(label.x.npc = .025) +
  # geom_hline(yintercept = min_covs, linetype=2) + 
  # geom_vline(xintercept = min_covs, linetype=2) + 
  scale_x_continuous(limits = c(0, 100)) +
  scale_y_continuous(limits = c(0, 100)) +
  labs(x="Query cov. BlastP",y="Query cov. Foldseek") +
  scale_fill_gradientn(colors=wespal) +
  coord_fixed(expand = 0) +
  theme(legend.position = "none", panel.grid = element_blank()) 
qcov_p <- add_marginal(df, qcov_p, "qcov_Blast", "qcov_Foldseek", "singleton", palette_singleton_p)


lddt_p <- df %>% 
  select(lddt, singleton) %>% 
  ggplot(aes(color=singleton, lddt)) + 
  geom_density() + 
  # scale_y_continuous(expand = c(0.01,0)) +
  # scale_x_continuous(limits = c(params$low_confidence, 100),
  #                    expand = c(0,0)) +
  scale_color_manual(values = palette_singleton_p) + 
  labs(x="LDDT") + 
  theme(legend.position = "none", 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

alntmscore_p <- df %>% 
  select(alntmscore, singleton) %>% 
  ggplot(aes(color=singleton, alntmscore)) + 
  geom_density() + 
  # coord_cartesian(expand = 0) +
  labs(x="TM-Score") + 
  scale_color_manual(values = palette_singleton_p)+ 
  theme(legend.position = "none", 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

plddt_p <- df %>% 
  select(target, singleton) %>% 
  mutate(target=gsub("AF-|-F1", "", target)) %>% 
  left_join(lddt_df, by="target") %>% 
  ggplot(aes(color=singleton, mean_lddt)) + 
  geom_density(bw=1) + 
  # scale_y_continuous(expand = c(0.01,0)) +
  scale_x_continuous(limits = c(params$low_confidence, 100)) +
  scale_color_manual(values = palette_singleton_p) +   
  labs(x="Mean target pLDDT") + 
  theme(legend.position = "none", 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

df_multi <- df_filt %>% 
  select(query, target, singleton) %>% 
    head(n = 1000) %>%
  left_join(CATH_df, by=c("query"="target")) %>% 
  filter(!is.na(Pfam), !is.na(Gene3D)) %>% 
  left_join(CATH_df, by="target") %>% 
  rowwise() %>% 
  mutate(Gene3D.x=strsplit(Gene3D.x, ";"),
         Pfam.x=strsplit(Pfam.x, ";"),
         Gene3D.y=strsplit(Gene3D.y, ";"),
         Pfam.y=strsplit(Pfam.y, ";")) %>% 
  mutate(n_3d_common=length(intersect(Gene3D.x,Gene3D.y)),
         n_Pfam_common=length(intersect(Pfam.x,Pfam.y))) %>% 
  # filter(n_Pfam_common>0, n_3d_common>0) %>% 
  select(-c(Pfam.x,Pfam.y,Gene3D.x,Gene3D.y)) %>% 
  mutate(Jaccard_3d=n_3d_common/pmax(n_3d_domains.x,n_3d_domains.y),
         Jaccard_Pfam=n_Pfam_common/pmax(n_Pfam_domains.x,n_Pfam_domains.y)) %>% 
  select(query, target, singleton, contains("Jaccard"), contains("common"))


CATH_plot <- df_multi %>% 
    pivot_longer(contains("Jaccard"), values_to = "Jaccard") %>% 
    mutate(name=gsub("Jaccard_", "", name),
           name=gsub("3d", "CATH", name),
           Jaccard=cut_width(Jaccard, .1, boundary = 0)) %>% 
  filter(!is.na(Jaccard)) %>% 
  group_by(name, singleton, Jaccard) %>% 
  summarise(n=n()) %>% 
  mutate(freq=n/sum(n)) %>%
  ggplot(aes(Jaccard, freq, fill=singleton)) + 
  geom_bar(stat = "identity",position = "dodge") + 
  facet_grid(name~.) +
  labs(y="% in Singleton group", x="Jaccard Index", 
       fill="Target sets") +
  scale_y_continuous(expand = expansion(0.01,0)) + 
  scale_fill_manual(values=palette_singleton_p) +
  theme(legend.position = "top",
        panel.background = element_rect(fill=NA, color="black"),
        axis.line = element_blank(),
        axis.text.x = element_text(size=8))


# prop_p <- group_by(df_filtered) %>% 
#   count(query, singleton) %>% 
#   group_by(query) %>% 
#   mutate(nn=sum(n), prop=n/nn) %>% 
#   ggplot(aes(prop, singleton, color=singleton)) + 
#   geom_boxplot() + 
#   labs(x="Proportion in tree", y="") +
#   scale_color_manual(values = palette_singleton_p) + 
#   theme(legend.position = "none")

fig2_first_row <- (pident_p | eval_p | qcov_p)
fig2_second_row <- ((lddt_p / alntmscore_p / plddt_p) | CATH_plot) + plot_layout(widths = c(1,2))
fig2 <-  (fig2_first_row / fig2_second_row) +
  plot_annotation(tag_levels = 'A') & 
  theme(plot.tag = element_text(size = 10, face = "bold", family = "Helvetica"))

ggsave("draft/figures/fig2.pdf", fig2, width = 10.5, height = 8)
```

## Fig3


```{r read_trees}
df_trees <- read_delim(glue(outdir,"/trees/", params_dataset$seed, "_unrooted_trees.txt"),
                       show_col_types = FALSE,
                       col_names = c("gene", "target", "alphabet", "model", "tree")) %>% 
  mutate(model=factor(model, levels=models)) %>% 
  filter(model!="FTPY")

# ML trees with selected model and LL
df_ml <- read_delim(glue(outdir,"/trees/", params_dataset$seed, "_mltrees.txt"), 
                    delim = "\t", 
                    show_col_types=FALSE) %>% 
  # mutate(fn=gsub("_mltrees.txt", "", basename(fn))) %>% 
  # separate(fn, c("gene", "target", "alphabet", "model"), sep = "_") %>% 
  mutate(model=factor(gsub("3DI", "3Di", 
                           gsub("GTR20", "GTR",
                                gsub("\\+.*", "", Model))), levels=models), 
         rate=sub("\\+", "", gsub("GTR20|3DI|LG", "", Model)))

disco_fls <- list.files(glue(outdir, "/reco/disco"), pattern = "output.nwk", full.names = T)

disco_rf <- NULL
for (file in disco_fls){
  a <- read.tree(file)
  rf <- TreeDist::RobinsonFoulds(sptree, a, normalize = T)
  ntips <- sapply(a, function(x) length(x$tip.label))
  disco_rf <- bind_rows(disco_rf, 
                        tibble(ntips=ntips, rf=rf, bn=rep(basename(file), length(rf))))
}


# Read all trees
ts <- read.tree(text = df_trees$tree)
names(ts) <- paste(df_trees$gene, df_trees$target, df_trees$alphabet, df_trees$model, sep = "_")

# Compare trees with RF distance
rf_df <- filter(df_trees, target=="common") %>% 
  select(-alphabet) %>% 
  inner_join(x=., y = ., by = c("gene", "target")) %>%
  filter(!is.na(tree.x), !is.na(tree.y)) %>%
  rowwise() %>%
  mutate(RF = phangorn::RF.dist(normalize = TRUE, 
                                read.tree(text = tree.x), read.tree(text=tree.y)),
         n_tips = length(read.tree(text = tree.x)$tip.label)) %>%
  select(-tree.x, -tree.y)

rf_text <- group_by(rf_df, model.x, model.y) %>% 
  # filter(as.integer(model.x)<=as.integer(model.y)) %>% 
  summarise(value=median(RF, na.rm=T))

# get BS distribution
df_bs <- fortify(ts) %>%
  filter(!isTip) %>%
  mutate(support=as.numeric(label)) %>%
  separate(.id, c("gene", "target", "alphabet", "model")) %>%
  filter(model!="QT") %>% 
  filter(!is.na(support)) %>% 
  select(gene, target, model, support)

n_tips <- sapply(ts, function(x) length(x$tip.label)) %>% 
  enframe() %>%
  separate(name, c("gene", "target", "alphabet", "model")) %>%
  rename("n_tips"="value")

n_sps <- sapply(ts, function(x) 
  length(unique(taxidmap[taxidmap$target %in% x$tip.label, ]$Tax_ID))) %>% 
  enframe() %>%
  separate(name, c("gene", "target", "alphabet", "model")) %>%
  rename("n_sps"="value")

var_df <- sapply(ts, function(x)
  var(adephylo::distRoot(phytools::midpoint_root(x)))) %>% 
  enframe() %>%
  separate(name, c("gene", "target", "alphabet", "model")) %>%
  rename("variance_r2t"="value")

# Reconciliation ranger+TCS scores
reco <- read_delim(glue(outdir,"/reco/", params_dataset$seed, "_notung.tsv"), 
                   show_col_types = FALSE, col_names = c("gene", "dups", "losses")) %>% 
  separate(gene, c("gene", "target", "alphabet", "model")) %>% 
  left_join(var_df) %>% 
  left_join(n_tips) %>% 
  left_join(n_sps) %>% 
  mutate(model=factor(model, levels=models),
         n_events = (dups+losses)/n_tips) %>% 
  rename("targets"="target")
```


```{r fig3}

plot_DL <- reco %>% 
  filter(model!="FTPY") %>% 
  ggplot(aes(model, n_events, fill=targets)) +
  geom_boxplot(outlier.size = .1) +
  labs(y="(D+L)/# Tips", x="",fill="Target sets") +
  # ggthemes::scale_fill_solarized() +
  scale_fill_manual(values = palettes_method) +
  theme(legend.position = "none")

plot_rf <- disco_rf %>% 
  filter(ntips>10) %>% 
  # filter(ntips>=length(sptree$tip.label)/2) %>%
  separate(bn, c("seed", "targets", "alphabet", "model")) %>% 
  mutate(model=factor(model, levels=models)) %>% 
  ggplot(aes(model, rf, fill=targets)) + 
  geom_boxplot() + 
  scale_fill_manual(values = palettes_method) +
  labs(x="", y="Norm. RF to species tree", fill="Target sets")


nodes_clades <- fortify(sptree) %>% 
  # filter(!isTip) %>% 
  mutate(ordered = rank(y)) %>% 
  mutate(label_int = fct_reorder(label, y)) %>%
  select(label_int, ordered)

apro_files <- list.files(glue(outdir, "/reco"), 
                         pattern = "*support.nwk", full.names = TRUE)
apro_trees <- read.tree(text = sapply(apro_files, readLines))
names(apro_trees) <- sapply(str_split(basename(apro_files), "_"), 
                            function(x) paste0(x[2], "_", x[4]))

trees_df <- fortify(apro_trees) 
nodes_df <- trees_df %>% 
  group_by(.id) %>% 
  mutate(ordered = rank(y)) %>% 
  left_join(nodes_clades) %>% 
  separate(.id, c("targets", "model")) %>% 
  filter(!isTip, label!="") %>% 
  mutate(label = gsub("\\[|\\]|\\'", "", str_replace_all(label, "[a-z]{1,2}[0-9]=", ""))) %>% 
  separate(label, c("pp1", "pp2", "pp3", "f1", "f2", "f3", "q1", "q2", "q3"), 
           ";", convert = TRUE) %>% 
  mutate(freq=f1+f2+f3, 
         model=factor(model, levels=models))

plot_apro <- nodes_df %>%
    filter(model!="FTPY") %>% 
  ggplot(aes(model, q1, fill=targets)) + 
  geom_boxplot() + 
  scale_fill_manual(values = palettes_method) +
  labs(x="", y="Quartet support", fill="Target sets")

# firstrow <- (plot_rf + plot_DL + 
#     plot_layout(guides = "collect"))
# 
# secondrow <- ( + 
#     plot_layout(guides = "collect", widths = c(1,2)))

fig3 <- plot_rf + plot_apro + plot_DL +
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 10, face = "bold", family = "Helvetica"),
        legend.position = "bottom")

ggsave("draft/figures/fig3.pdf", fig3, width = 10, height = 4)
# reco_rnkgs_plot <- reco %>% 
#   group_by(gene, target) %>% 
#   slice_min(n_events, with_ties = T) %>% 
#   group_by(target, model) %>% 
#   count() %>% 
#   ggplot(aes(target, n, fill=model)) + 
#   geom_bar(stat = "identity") +
#   geom_text(aes(label=n), position = position_stack()) +
#   scale_fill_manual(values = palettes_model)
```


```{r FigS1}
plot_bs <- df_bs %>% 
  mutate(targets=target,
         model=factor(model, levels=models)) %>% 
  ggplot(aes(model, support, fill=targets)) +
  geom_boxplot(outlier.size = .1) +
  scale_fill_manual(values = palettes_method) +
  theme_classic() +
  labs(y="Boostrap", x="", fill="") +
  theme(legend.position = "bottom")

var_r2t <- reco %>% 
  filter(model!="FTPY", targets=="common") %>% 
  ggplot(aes(variance_r2t, model)) +
  geom_violin(scale = "width", color="black", fill="grey40") +
  scale_x_continuous(limits = c(0,.5), expand = expansion(mult=0, add = c(0,0.05))) +
  labs(x="Variance Root-to-tip distance") +
  # scale_fill_manual(values = palettes_model) +
  theme(legend.position = 'none',
        legend.margin=margin(0,0,0,0),
        legend.box.margin=margin(-10,0,0,0))

rf_plot <- rf_text %>% 
  ggplot(aes(model.x, model.y)) +
  geom_tile(aes( fill=value)) +
  geom_text(aes(label=round(value, 2)), color="white") +
  scale_fill_distiller(palette="YlGnBu", direction = -1, limits = c(0,1), 
                       name="Median RF") + 
  labs(x="Median RF Distance", y="") +
  theme_classic() +
  theme(legend.position = "none", 
        axis.line = element_blank(),
        axis.ticks = element_blank())

df_red <- select(df_filtered, query, target, singleton) 

res <- NULL
for (idx in 1:nrow(df_trees)){
  t <- ts[[idx]]
  seed <-  t$tip.label[grepl(df_trees$gene[idx], t$tip.label)]
  nn <- as.matrix(adephylo::distTips(t, method = "nNodes"))[seed, ]
  max_dist <- max(nn)

  row <- enframe(nn, name = "target", value="nn_dist") %>% 
    mutate(query=seed, 
           max_dist=max_dist,
           targets=df_trees$target[idx],
           alphabet=df_trees$alphabet[idx],
           model=df_trees$model[idx]) %>% 
    left_join(df_red, by=c("query", "target"))
  res <- bind_rows(res, row)
}

# for what this should be grouped?
plot_dist <- res %>% 
  filter(query!=target) %>% 
  group_by(query, singleton, model) %>%
  summarise(median_dist=median(nn_dist/max_dist), n=n()) %>% 
  ggplot(aes(median_dist, model,  fill=singleton)) + 
  ggdist::stat_halfeye(alpha=.8, point_colour = NA, .width = 0, height=.6, 
                       position = position_nudge(y=.15)) +
  geom_boxplot(
    width=.2,
    position = position_dodge(),
    # removing outliers
    outlier.color = NA,
    alpha = 1,
    # position= position_nudge(y=-.5)
  ) + 
  labs(x="Normalised internode distance", y="Method", fill="") +
  scale_x_continuous(expand = c(0,0.01), limits = c(0,1)) +
  theme(panel.border = element_rect(fill = NA)) + 
  scale_fill_manual(values = palette_singleton_p) + 
  theme(legend.position = "bottom")

fig_s1 <- ((plot_bs / rf_plot) + 
             plot_layout(heights = c(2,1)) | 
             (plot_dist / var_r2t) + 
             plot_layout(heights = c(2,1))) +
  plot_layout(widths = c(1, 2)) +
  plot_annotation(tag_levels = 'A') & 
  theme(plot.tag = element_text(size = 10, face = "bold", family = "Helvetica"))
ggsave("draft/figures/figS1.pdf", fig_s1, width = 10, height = 6)
```

```{r FigS2}
pp1_plot <- nodes_df %>%
    filter(model!="FTPY") %>% 
    ggplot(aes(q1, label_int, fill=model, color=model, group=model)) +
    geom_linerange(aes(x=x, xmin=xmin, xmax=xmax, color=model),
                   data = . %>% 
                     group_by(model, label_int) %>% 
                     summarise(x=mean(q1), xmin=min(q1), xmax=max(q1)),
                   position = position_dodge(width=.7), color="grey70") +
    geom_point(aes(shape=targets),
               position = position_dodge(width=.7)) +
  scale_fill_manual(values=palettes_model) +
  scale_color_manual(values = palettes_model_d) +
  scale_shape_manual(values = c(21,22,23,24)) +
  labs(y="Node") +
  guides(color = guide_legend(override.aes = list(pch = 21, size=3)),
         shape = guide_legend(override.aes = list(size=3))) +
  theme(legend.position = "bottom", legend.box="vertical") + 
  theme_bw()

labeled_sptree <- sptree
labeled_sptree$tip.label <- tax$Species[match(labeled_sptree$tip.label, tax$mnemo)]

sptree_plot <- ggtree(labeled_sptree) + 
  geom_tiplab(aes(label=sub("[^A-Z].+? ", ". ", label)),fontface="italic",
              align = TRUE) + 
  geom_nodelab(color="firebrick", 
               nudge_x = -.15, hjust = 1, 
               nudge_y = .3, size=3.5)+
  xlim(c(-4, 21))

fig_s2 <- (sptree_plot + pp1_plot) +
  plot_layout(widths = c(1, 1.5)) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 10, face = "bold", family = "Helvetica"))
ggsave("draft/figures/figS2.pdf", fig_s2, width = 12, height = 6, device = cairo_pdf)
```

