---
title: "Structural analysis"
# knit: (function(input_file, encoding) {
#   rmarkdown::render(input_file,
date: "`r format(Sys.time(), '%d-%b-%Y')`"
author: "Giacomo Mutti"
output:
    rmdformats::html_clean:
        number_sections: false
        code_folding: hide
        self_contained: true
        fig_caption: true
        # thumbnails: false
        # lightbox: false
        gallery: true
        toc_depth: 3
        highlight: kate
---

```{r setup, include=FALSE, dev="CairoPNG"}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE, cache.lazy = FALSE,
                      crop = TRUE, dev="CairoPNG")
knitr::knit_hooks$set(crop = knitr::hook_pdfcrop)
knitr::opts_knit$set(root.dir = "../../")
```

```{r libraries, message=FALSE, warning=FALSE, cache=FALSE}
library(tidyverse)
library(patchwork)
library(ggtree)
library(cowplot)
library(glue)
library(ggh4x)
source("workflow/scripts/functions.R")
theme_set(theme_classic(base_family = "Helvetica"))

models <- names(palettes_model)

params <- yaml::read_yaml("config/params.yaml")
params_dataset <- yaml::read_yaml("config/Hsap_draft.yaml")

homodir <- glue(params$outdir, "homology/", params_dataset$homology_dataset)
phylodir <- glue(params$outdir, "phylogeny/", params_dataset$phylo_dataset)
```

```{r read_stuff, cache=FALSE}
seeds <- readLines(params_dataset$test_seeds)

tax <- read_delim(glue("results/data/meta/",
                       params_dataset$homology_dataset, "_uniprot_genomes.tsv"), 
                  show_col_types=F)
taxmap <- read_delim(glue(homodir, "/db/taxidmap"), 
                     delim = "\t", col_names = c("label", "Tax_ID"),
                     show_col_types = FALSE) %>% 
  left_join(tax, by = "Tax_ID") %>% 
  dplyr::select(label, Proteome_ID, mnemo)

taxidmap <- read_delim(glue(homodir,"/db/taxidmap"), 
                       col_names = c("target", "Tax_ID"), show_col_types=F)

sptree <- read.tree(params_dataset$species_tree)
translate <- deframe(distinct(dplyr::select(taxmap, Proteome_ID, mnemo)))
# sptree$tip.label <- names(translate)[match(translate, sptree$tip.label)]

species_dist_df <- tibble()
for (sp in sptree$tip.label){
  if (sp=="Hsap") {next}
  dist <- castor::get_all_distances_to_tip(sptree, "Hsap")[MRCA(sptree, "Hsap", sp)]
  mrca <- c(sptree$tip.label, sptree$node.label)[MRCA(sptree, "Hsap", sp)]
  species_dist_df <- bind_rows(species_dist_df, tibble(mnemo=sp, value=dist, mrca=mrca))
}

table_columns <- c('Proteome_ID', 'Tax_ID', 'count1', 'count2', 'count3', 
                   'genome', 'source', 'species', 'mnemo')

lddt_df <- read_delim(paste0(params$data_dir, "structures/", 
                             tax$Proteome_ID, "/", tax$Proteome_ID, "_mean_plddt.tsv"),
           col_names = c("target", "mean_lddt"), show_col_types = FALSE)

aln_seeds <- readLines(glue(phylodir,"/ids/", params_dataset$seed, "_aln.ids"))
```

# Homology

```{r read_self_blast}
blast_self <- read_delim(glue(homodir, "/allvall/",
                              params_dataset$seed, "_", params_dataset$seed, "_blast.tsv"),
                         show_col_types=FALSE,
                         col_names = blast_columns) %>% 
  # filter(query %in% seeds) %>%
  filter(query==target) %>% 
  mutate(blast_selfbit=bitscore) %>% 
  select(query, blast_selfbit)

fs_self <- read_delim(glue(homodir, "/allvall/", params_dataset$seed, 
                           "_", params_dataset$seed, "_fs.tsv"), 
                      show_col_types = FALSE,
                         col_names = blast_columns) %>% 
  # filter(query %in% seeds) %>%
  filter(query==target) %>% 
  mutate(fs_selfbit=bitscore) %>% 
  select(query, fs_selfbit)

self <- full_join(blast_self, fs_self, by="query")
```


```{r read_homo}
fs <- read_delim(glue(homodir, "/", params_dataset$seed, "_fs.tsv"), 
                 col_names = fs_columns, show_col_types = FALSE) %>% 
  # filter(query %in% seeds) %>%
  mutate(evalue = ifelse(evalue==0, 1e-180, evalue)) %>%
  left_join(taxidmap, by = "target")

blast <- read_delim(glue(homodir, "/", params_dataset$seed, "_blast.tsv"), 
                    col_names = blast_columns, show_col_types = FALSE) %>% 
  # filter(query %in% seeds) %>%
  mutate(evalue = ifelse(evalue==0, 1e-180, evalue))
```

```{r merge_homo}
df <- full_join(blast, fs,
                by = c("query", "target"), suffix=c("_Blast", "_Foldseek")) %>%
    filter(query!=target) %>% 
  left_join(self, by="query") %>%
  mutate(pident_Foldseek = 100*pident_Foldseek, 
         qcov_Foldseek = 100*qcov_Foldseek,
         singleton = case_when(is.na(evalue_Foldseek) ~ "only Blast",
                               is.na(evalue_Blast) ~ "only Foldseek", 
                               .default = "Common")) 


blast_filt <- filter(blast, 
                         length/qlen*100>params$coverage, 
                         length/slen*100>params$coverage,
                         evalue<as.numeric(params$eval_both)) %>% 
  group_by(query) %>% 
  slice_head(n = params$max_seqs) %>% 
  ungroup()

fs_filt <- filter(fs, 
                      qcov*100>params$coverage, 
                      tcov*100>params$coverage, 
                      evalue<as.numeric(params$eval_both)) %>% 
  group_by(query) %>% 
  slice_head(n = params$max_seqs) %>% 
  ungroup()

df_filt <- full_join(blast_filt, fs_filt,
                by = c("query", "target"), suffix=c("_Blast", "_Foldseek")) %>%
    filter(query!=target) %>% 
  left_join(self, by="query") %>%
  mutate(pident_Foldseek = 100*pident_Foldseek, 
         qcov_Foldseek = 100*qcov_Foldseek,
         singleton = case_when(is.na(evalue_Foldseek) ~ "only Blast",
                               is.na(evalue_Blast) ~ "only Foldseek", 
                               .default = "Common"))

df %>% 
  group_by(singleton) %>% 
  summarise(n(), n()/nrow(df))

df_filt %>% 
  group_by(singleton) %>% 
  summarise(n(), n()/nrow(df_filt))
```


```{r cath}
CATH_df <- read_delim(paste0(params$data_dir, "cath/", tax$Proteome_ID, "_cath.tsv"),
                      delim = "\t", id = "proteome", show_col_types = FALSE) %>%
  mutate(proteome = gsub("_cath.tsv", "", basename(proteome))) %>%
         # n_3d_domains = str_count(Gene3D, ";"),
         # n_Pfam_domains = str_count(Pfam, ";")
  mutate(target=paste0("AF-", Entry, "-F1")) %>% 
  rename("CATH"="Gene3D") %>% 
  select(target, Pfam, CATH, -proteome, -Entry) %>% 
  mutate(CAT = gsub("(\\d+\\.\\d+\\.\\d+)\\.\\d+", "\\1", CATH),
         CA = gsub("(\\d+\\.\\d+)\\.\\d+", "\\1", CAT),
         C = gsub("\\.\\d+", "\\1", CA))

threed_df <- read_delim(paste0(params$data_dir, "ids/", tax$Proteome_ID, "_3ds.tsv"),
           delim = "\t", id = "proteome", show_col_types = FALSE) %>% 
  mutate(proteome = gsub("_ids.tsv", "", basename(proteome))) %>% 
  mutate(target=paste0("AF-", Entry, "-F1")) %>% 
  select(-proteome, -Entry) %>% 
  left_join(taxmap, by=c("target"="label")) %>% 
  mutate(sp = gsub("..*_", "", `Entry Name`))

ortho_dbs <- read_delim(paste0(params$data_dir, "ids/", tax$Proteome_ID, "_ids.tsv"),
                      delim = "\t", id = "proteome", show_col_types = FALSE) %>% 
  mutate(proteome = gsub("_ids.tsv", "", basename(proteome))) %>% 
  mutate(target=paste0("AF-", Entry, "-F1")) %>% 
  select(-proteome, -Entry)

ortho_dbs <- full_join(CATH_df, ortho_dbs)
```

## Fig1

```{r fig1}
fig1a <- df %>%
  group_by(singleton) %>%
  summarise(n = n()) %>%
  ungroup() %>%
  mutate(prop=n/sum(n), Filtering="Before") %>%
  bind_rows(df_filt %>%
            group_by(singleton) %>%
            summarise(n = n()) %>%
            ungroup() %>%
            mutate(prop=n/sum(n), Filtering="After")) %>%
  mutate(Filtering=factor(Filtering, labels = c("After", "Before"), ordered = T),
         singleton=case_when(singleton=="only Blast"~"Only Bp", 
                             singleton=="only Foldseek" ~ "Only Fs",
                             .default = singleton)) %>%
  ggplot(aes(singleton, n, fill=singleton)) +
  geom_bar(aes(alpha=Filtering), stat = "identity", position = position_dodge2(reverse=TRUE)) +
  scale_fill_manual(values = palette_singleton_s, guide="none") +
  scale_alpha_manual(values = c(1,.6)) +
  scale_y_continuous(expand = c(0,0), labels = scales::label_number(scale = 1e-6)) +
  labs(fill="Class", x="Class", y="Milions of aligned protein pairs") +
  theme(legend.position = "bottom", legend.box="vertical")

fig1b <- lddt_df %>% 
  ggplot(aes(mean_lddt)) + 
  geom_density(fill=NA, size=1) +
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  labs(x="Mean pLDDT", y="") +
  geom_vline(xintercept = 40, linetype=2) +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank(),
        axis.line.y = element_blank())

fig1 <- (fig1a | fig1b) + plot_layout(widths = c(1,2))

ggsave("draft/figures/Fig1_tmp.pdf", fig1, width = 8, height = 4)
```

## Fig2

```{r fig2}
eval_p <- ggplot(df, aes(-log10(evalue_Blast), -log10(evalue_Foldseek))) + 
  stat_bin_2d(breaks=seq(0,20,0.4)) + 
  scale_fill_gradientn(colors=wespal) +
  ggpubr::stat_cor(label.x.npc = .025) +
  coord_cartesian(expand = 0) +
  labs(x="-log10(Evalue BlastP)",y="-log10(Evalue Foldseek)") +
  scale_x_continuous(limits = c(0, 20)) +
  scale_y_continuous(limits = c(0, 20)) +
  theme(legend.position = "none", panel.grid = element_blank()) 
eval_p <- add_marginal(mutate(df, 
                              evalue_Blast=-log10(evalue_Blast), 
                              evalue_Foldseek=-log10(evalue_Foldseek)), 
                       eval_p, "evalue_Blast", "evalue_Foldseek", "singleton", palette_singleton_p)

pident_p <- ggplot(df, aes(pident_Blast, pident_Foldseek)) + 
  stat_bin_2d(breaks=seq(0,100,2)) + 
  scale_fill_gradientn(colors=wespal) +
  ggpubr::stat_cor(label.x.npc = .025) +
  coord_cartesian(expand = 0) +
  scale_x_continuous(limits = c(0, 100)) +
  scale_y_continuous(limits = c(0, 100)) +
  labs(x="% Identity BlastP",y="% Identity Foldseek") +
  theme(legend.position = "none", panel.grid = element_blank())
pident_p <- add_marginal(df, pident_p, "pident_Blast", "pident_Foldseek", "singleton", palette_singleton_p)


qcov_p <- ggplot(df, aes(qcov_Blast, qcov_Foldseek)) + 
  stat_bin_2d(breaks=seq(0,100,2)) + 
  ggpubr::stat_cor(label.x.npc = .025) +
  scale_x_continuous(limits = c(0, 100)) +
  scale_y_continuous(limits = c(0, 100)) +
  labs(x="Query cov. BlastP",y="Query cov. Foldseek") +
  scale_fill_gradientn(colors=wespal) +
  coord_cartesian(expand = 0) +
  theme(legend.position = "none", panel.grid = element_blank()) 
qcov_p <- add_marginal(df, qcov_p, "qcov_Blast", "qcov_Foldseek", "singleton", palette_singleton_p)

lddt_p <- df %>% 
  select(lddt, singleton) %>% 
  ggplot(aes(color=singleton, lddt)) + 
  geom_density() + 
  scale_y_continuous(expand = expansion(c(0.02,0.02))) +
  scale_x_continuous(limits = c(0, 1), 
                     expand = expansion(c(0.02,0.02))) +
  scale_color_manual(values = palette_singleton_p, guide=FALSE) + 
  labs(x="LDDT", color = "Target sets") + 
  theme(legend.position = "none", 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

alntmscore_p <- df %>% 
  select(alntmscore, singleton) %>% 
  ggplot(aes(color=singleton, alntmscore)) + 
  geom_density() + 
  scale_y_continuous(expand = expansion(c(0.02,0.02))) +
  scale_x_continuous(limits = c(0, 1), 
                     expand = expansion(c(0.02,0.02))) +
  labs(x="TM-Score", color = "Target sets") + 
  scale_color_manual(values = palette_singleton_p, guide=FALSE)+ 
  theme(legend.position = "none", 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

plddt_p <- df %>% 
  select(target, singleton) %>% 
  mutate(target=gsub("AF-|-F1", "", target)) %>% 
  left_join(lddt_df, by="target") %>% 
  ggplot(aes(color=singleton, mean_lddt)) + 
  geom_density(bw=1) + 
  scale_y_continuous(expand = expansion(c(0.02,0.02))) +
  scale_x_continuous(limits = c(params$low_confidence, 100), 
                     expand = expansion(c(0.02,0.02))) +
  scale_color_manual(values = palette_singleton_p, guide=FALSE) +   
  labs(x="Mean target pLDDT", color = "Target sets") + 
  theme(legend.position = "none", 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

dbs <- colnames(ortho_dbs)[2:dim(ortho_dbs)[2]]
dbs <- setdiff(dbs, c("InParanoid"))
dbs <- dbs[2:5]

binwidth <- 0.1

# Loop through each column name and apply a function
# jaccard_df_mono <- map_dfr(dbs, ~ {
#   col_name <- .x
#   select(df_filt, query, target, singleton) %>% 
#     left_join(select(ortho_dbs, target, col_name), by=c("query"="target")) %>% 
#     left_join(select(ortho_dbs, target, col_name), by=c("target")) %>% 
#     rename("query_set"=paste0(col_name, ".x"), "target_set"=paste0(col_name, ".y")) %>% 
#     mutate(n_domain_q = str_count(pattern = "\\;", query_set)) %>% 
#     # Get only monodomain queries
#     filter(n_domain_q==1) %>% 
#     filter(!is.na(query_set), !is.na(target_set)) %>% 
#     mutate(
#       jaccard = map2_dbl(strsplit(query_set, ";"), strsplit(target_set, ";"),
#                          ~ length(intersect(.x, .y)) / length(union(.x, .y)))
#     ) %>% 
#     group_by(query, singleton) %>% 
#     summarise(avg_Jaccard = mean(jaccard))  %>% 
#     mutate(db=col_name)
# })
# 
# jaccard_df_NA <- map_dfr(dbs, ~ {
#   col_name <- .x
#   select(df_filt, query, target, singleton) %>%
#     left_join(select(ortho_dbs, target, col_name), by=c("query"="target")) %>%
#     left_join(select(ortho_dbs, target, col_name), by=c("target")) %>%
#     rename("query_set"=paste0(col_name, ".x"), "target_set"=paste0(col_name, ".y")) %>%
#     # If target has na no we count mismatch
#     filter(!is.na(query_set)) %>%
#     mutate(
#       jaccard = map2_dbl(strsplit(query_set, ";"), strsplit(target_set, ";"),
#                          ~ length(intersect(.x, .y)) / length(union(.x, .y)))
#     )  %>%
#     group_by(query, singleton) %>%
#     summarise(avg_Jaccard = mean(jaccard)) %>%
#     mutate(db=col_name)
# 
# })


jaccard_df <- map_dfr(dbs, ~ {
  col_name <- .x
  select(df_filt, query, target, singleton) %>% 
    left_join(select(ortho_dbs, target, col_name), by=c("query"="target")) %>% 
    left_join(select(ortho_dbs, target, col_name), by=c("target")) %>% 
    rename("query_set"=paste0(col_name, ".x"), "target_set"=paste0(col_name, ".y")) %>% 
    # If any of query or target has na no we dont count
    filter(!is.na(query_set), !is.na(target_set)) %>% 
    mutate(
      jaccard = map2_dbl(strsplit(query_set, ";"), strsplit(target_set, ";"),
                         ~ length(intersect(.x, .y)) / length(union(.x, .y)))
    )  %>% 
    group_by(query, singleton) %>% 
    summarise(avg_Jaccard = mean(jaccard)) %>% 
    mutate(db=col_name) 

})

p_dbs <- ggplot(jaccard_df, aes(avg_Jaccard, 1, color=singleton, alpha=db)) +
  stat_ecdf(pad=FALSE) +
  scale_color_manual(values=palette_singleton_p) +
  # scale_linetype_manual(values = c(2,2,2,2,1)) +
  scale_alpha_manual(values = seq(0.3, 1, length.out=4)) +
  coord_fixed() +
  scale_y_continuous(expand = expansion(0.0,0)) +  
  scale_x_continuous(expand = expansion(add = c(0,0.01))) +  
  guides(colour = guide_legend(order = 1),  
         alpha = guide_legend(title = "CATH Level", order = 2)) +
  theme(panel.border = element_rect(fill = NA), 
        legend.position = "bottom",
        panel.grid.major = element_line(linetype = 2)) + 
  # theme_bw(base_family = "Aptos") +
  labs(x="Mean Jaccard Index", y="ECDF", 
       alpha="DB", linetype="DB", color="Target sets")

# p_dbs_mono <-  ggplot(jaccard_df_mono, aes(avg_Jaccard, 1, color=singleton, 
#                                 linetype=db, alpha=db)) +
#   stat_ecdf(pad=FALSE) +
#   scale_color_manual(values=palette_singleton_p) +
#   scale_linetype_manual(values = c(2,2,2,2,1)) +
#   scale_alpha_manual(values = c(0.4,0.6,0.8,1,1)) +
#   coord_fixed() +
#   scale_y_continuous(expand = expansion(0.02,0)) +  
#   scale_x_continuous(expand = expansion(0.02,0)) +  
#   theme_bw(base_family = "Aptos") +
#   labs(title="Only Monodomain queries", x="Mean Jaccard Index", y="ECDF", 
#        alpha="DB", linetype="DB", color="Target sets")
# 
# p_dbs_NA <-  ggplot(jaccard_df_NA, aes(avg_Jaccard, 1, color=singleton,
#                                 linetype=db, alpha=db)) +
#   stat_ecdf(pad=FALSE) +
#   scale_color_manual(values=palette_singleton_p) +
#   scale_linetype_manual(values = c(2,2,2,2,1)) +
#   scale_alpha_manual(values = c(0.4,0.6,0.8,1,1)) +
#   coord_fixed() +
#   scale_y_continuous(expand = expansion(0.02,0)) +
#   scale_x_continuous(expand = expansion(0.02,0)) +
#   theme_bw(base_family = "Aptos") +
#     labs(title="NA=mismatch", x="Mean Jaccard Index", y="ECDF", 
#        alpha="DB", linetype="DB", color="Target sets")

# jaccard_plot <- (p_dbs_NA | p_dbs_mono) + 
#   plot_layout(guides = "collect", 
#               axis_titles = "collect") & theme(legend.position = 'bottom')

fig2_first_row <- (pident_p | eval_p | qcov_p)
fig2_second_row <- ((lddt_p / alntmscore_p / plddt_p) | p_dbs) + 
  plot_layout(guides = "collect") & 
  theme(legend.position = 'top')
fig2 <- (fig2_first_row / fig2_second_row) +
  plot_layout(heights = c(1,1.3)) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 10, face = "bold", family = "Aptos"))

ggsave("draft/figures/Fig2.pdf", fig2, width = 8, height = 7, device = cairo_pdf)
# ggsave("draft/figures/Fig_Supp_CATH.pdf", jaccard_plot, width = 8, height = 4, device = cairo_pdf)
```

## Fig3


```{r read_trees}
models_draft <- setdiff(models, c("QT", "FTPY", "AF", "LLM"))
df_trees <- read_delim("test/flashforward/out/unrooted_trees.txt",
                       show_col_types = FALSE,
                       col_names = c("gene", "target", "alphabet", "model", "tree")) %>% 
  mutate(model=factor(model, levels=models)) %>% 
  filter(!is.na(tree), model %in% models_draft) %>% 
  # some FM trees have negative branch lengths
  mutate(tree = gsub(":-", ":", tree)) 
# 722*8*4=23104 trees

# ML trees with selected model and LL
df_ml <- read_delim("draft/mltrees.txt", 
                    delim = "\t", 
                    show_col_types=FALSE) %>% 
  # mutate(fn=gsub("_mltrees.txt", "", basename(fn))) %>% 
  # separate(fn, c("gene", "target", "alphabet", "model"), sep = "_") %>% 
  mutate(model=factor(gsub("3DI", "3Di", 
                           gsub("GTR20", "GTR",
                                gsub("\\+.*", "", Model))), levels=models), 
         rate=sub("\\+", "", gsub("GTR20|3DI|LG", "", Model)))

disco_fls <- list.files("test/flashforward/reco/", pattern = "^disco", full.names = T)

# Read all trees
ts <- read.tree(text = df_trees$tree)
names(ts) <- paste(df_trees$gene, df_trees$target, df_trees$alphabet, df_trees$model, sep = "_")

disco_rf <- get_disco_rf(disco_fls, sptree)
df_bs <- get_bs_df(ts)

n_tips <- sapply(ts, function(x) length(x$tip.label)) %>% 
  enframe() %>%
  separate(name, c("gene", "target", "alphabet", "model"), sep = "_") %>%
  rename("n_tips"="value")

var_df <- sapply(ts, function(x) { 
                 a <- phytools::midpoint_root(x) 
                 var(ape::dist.nodes(a)[a$edge[1,][1],1:length(a$tip.label)])
                 }) %>% 
  enframe() %>%
  separate(name, c("gene", "target", "alphabet", "model"), sep="_") %>%
  rename("variance_r2t"="value")

# Reconciliation ranger+TCS scores
# glue("results/phylogeny/", params_dataset$phylo_dataset, "/reco/", params_dataset$seed, "_notung.tsv")
reco <- read_delim("test/flashforward/out/notung.txt", 
                   show_col_types = FALSE, 
                   col_names = c("gene", "target", "alphabet", "model", "dups", "losses")) %>% 
  left_join(n_tips) %>% 
  left_join(var_df) %>% 
  mutate(model=factor(model, levels=models),
         n_events = (dups+losses)/n_tips) %>% 
  filter(model %in% models_draft) %>% 
  rename("targets"="target")
```


```{r fig3}
xax <- ifelse(df_model$algorithm=="Distance", "italic", "plain")

plot_DL <- reco %>% 
  left_join(df_model) %>% 
  ggplot(aes(weave_factors(model, data), n_events, 
             color=algorithm, fill=targets)) + 
  labs(y="(D+L)/# Tips", x="",fill="Target sets") +
  geom_boxplot(outlier.size = .1) +
  scale_color_manual(values = c("black", "black"),
                     labels = c("ML", expression(italic("Distance")))) +
  guides(x = "axis_nested", 
         colour=guide_legend(title = "Method", override.aes=list(linetype = 0, size=0))) +
  scale_fill_manual(values = palettes_method) +
    theme(
    axis.text.x = element_text(face = xax),
    ggh4x.axis.nestline.x = element_line(color="grey"),
    ggh4x.axis.nesttext.x = element_text(colour = "grey30", 
                                         face = "bold", size=10)
  )


plot_rf <- disco_rf %>% 
  mutate(model=factor(model, levels=models)) %>% 
  filter(nsps>length(sptree$tip.label)/2) %>%
  # separate(bn, c("seed", "targets", "alphabet", "model")) %>% 
  left_join(df_model) %>% 
  ggplot(aes(weave_factors(model, data), rf, 
             color=algorithm, fill=target)) + 
  labs(x="", y="Norm. RF to species tree", fill="Target sets") +
  geom_boxplot(outlier.size = .1) +
  scale_color_manual(values = c("black", "black"),
                     labels = c("ML", expression(italic("Distance")))) +
  guides(x = "axis_nested", 
         colour=guide_legend(title = "Method", override.aes=list(linetype = 0, size=0))) +
  scale_fill_manual(values = palettes_method) +
    theme(
    axis.text.x = element_text(face = xax),
    ggh4x.axis.nestline.x = element_line(color="grey"),
    ggh4x.axis.nesttext.x = element_text(colour = "grey30", 
                                         face = "bold", size=10)
  )

apro_files <- list.files("test/flashforward/reco/", 
                         pattern = "^apro*", full.names = TRUE)

apro_trees <- read.tree(text = sapply(apro_files, readLines))
names(apro_trees) <- sapply(str_split(basename(apro_files), "_"), 
                            function(x) paste0(x[2], "_", x[4]))

trees_df <- get_apro_stats(apro_trees) 

plot_apro <- trees_df %>% 
  left_join(df_model) %>% 
  ggplot(aes(weave_factors(model, data), q1, 
             color=algorithm,fill=targets)) + 
  labs(x="", y="Quartet support", fill="Target sets") + 
  geom_boxplot(outlier.size = .1) +
  scale_color_manual(values = c("black", "black"),
                     labels = c("ML", expression(italic("Distance")))) +
  guides(x = "axis_nested", 
         colour=guide_legend(title = "Method", 
                             override.aes=list(linetype = 0, size=0))) +
  scale_fill_manual(values = palettes_method) +
    theme(
    axis.text.x = element_text(face = xax),
    ggh4x.axis.nestline.x = element_line(color="grey"),
    ggh4x.axis.nesttext.x = element_text(colour = "grey30", 
                                         face = "bold", size=10)
  )

fig3 <- plot_rf + plot_apro + plot_DL +
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 10, face = "bold", family = "Helvetica"),
        legend.position="bottom",
        legend.margin=margin(0,0,0,0),
        legend.box.margin=margin(-25,0,-10,0)) 


ggsave("draft/figures/Fig3.pdf", fig3, width = 9, height = 2.8, device = cairo_pdf)
```


```{r FigS2}
# Compare trees with RF distance
rf_df <- get_rf_df(df_trees)
rf_text <- group_by(rf_df, model.x, model.y) %>% 
  # filter(as.integer(model.x)<=as.integer(model.y)) %>% 
  summarise(value=mean(RF, na.rm=T))

plot_bs <- df_bs %>% 
  filter(model!="FM") %>% 
  mutate(targets=target,
         model=factor(model, levels=models)) %>% 
  ggplot(aes(model, support, fill=targets)) +
  geom_boxplot(outlier.size = .1) +
  scale_fill_manual(values = palettes_method) +
  theme_classic() +
  labs(y="Boostrap", x="", fill="") +
  theme(legend.position = "bottom")

var_r2t <- reco %>% 
  filter(targets=="common", model %in% models_draft) %>% 
  ggplot(aes(variance_r2t, fct_reorder(model, variance_r2t))) +
  geom_violin(scale = "width", color="black", fill="grey40") +
  scale_x_continuous(limits = c(0,.5), expand = expansion(mult=0, add = c(0,0.05))) +
  labs(x="Variance Root-to-tip distance", y="") +
  # scale_fill_manual(values = palettes_model) +
  theme(legend.position = 'none',
        legend.margin=margin(0,0,0,0),
        legend.box.margin=margin(-10,0,0,0))

rf_plot <- rf_text %>% 
  ggplot(aes(model.x, model.y)) +
  geom_tile(aes(fill=value)) +
  geom_text(aes(label=round(value, 2)), color="white") +
  scale_fill_distiller(palette="YlGnBu", direction = -1, limits = c(0,1), 
                       name="Mean RF") + 
  labs(x="Mean RF Distance", y="") +
  theme_classic() +
  theme(legend.position = "none", 
        axis.line = element_blank(),
        axis.ticks = element_blank())

union_trees <- df_trees %>% 
  filter(target=="union", model %in% models_draft)
df_dist <- distance_to_seed(union_trees) 

plot_dist <- df_dist %>% 
  left_join(select(df_filt, query, target, singleton), 
            by = c("query", "target")) %>%
  filter(query!=target, !is.na(singleton)) %>% 
  group_by(query) %>% 
  mutate(max_dist = max(nn_dist)) %>% 
  group_by(query, singleton, model) %>%
  summarise(median_dist=median(nn_dist/max_dist), n=n()) %>% 
  ggplot(aes(median_dist, model,  fill=singleton)) + 
  # ggdist::stat_halfeye(alpha=.8, point_colour = NA, .width = 0, height=.6, 
  #                      position = position_nudge(y=.15)) +
  geom_boxplot(
    # width=.2,
    # position = position_dodge(),
    # removing outliers
    outlier.color = NA,
    alpha = 1,
    # position= position_nudge(y=-.5)
  ) + 
  labs(x="Normalised internode distance", y="Method", fill="") +
  scale_x_continuous(expand = c(0,0.01), limits = c(0,1)) +
  theme(panel.border = element_rect(fill = NA)) + 
  scale_fill_manual(values = palette_singleton_p) + 
  theme(legend.position = "bottom")

fig_s2 <- ((plot_bs / rf_plot) + 
             plot_layout(heights = c(2,1)) | 
             (plot_dist / var_r2t) + 
             plot_layout(heights = c(2,1))) +
  plot_layout(widths = c(1, 2)) +
  plot_annotation(tag_levels = 'A') & 
  theme(plot.tag = element_text(size = 10, face = "bold", family = "Helvetica"))

ggsave("draft/figures/Fig_S2.pdf", fig_s2, width = 10, height = 6)
```

# ```{r FigS2}
# pp1_plot <- nodes_df %>%
#     filter(model!="FTPY") %>% 
#     ggplot(aes(q1, label_int, fill=model, color=model, group=model)) +
#     geom_linerange(aes(x=x, xmin=xmin, xmax=xmax, color=model),
#                    data = . %>% 
#                      group_by(model, label_int) %>% 
#                      summarise(x=mean(q1), xmin=min(q1), xmax=max(q1)),
#                    position = position_dodge(width=.7), color="grey70") +
#     geom_point(aes(shape=targets),
#                position = position_dodge(width=.7)) +
#   scale_fill_manual(values=palettes_model) +
#   scale_color_manual(values = palettes_model_d) +
#   scale_shape_manual(values = c(21,22,23,24)) +
#   labs(y="Node") +
#   guides(color = guide_legend(override.aes = list(pch = 21, size=3)),
#          shape = guide_legend(override.aes = list(size=3))) +
#   theme(legend.position = "bottom", legend.box="vertical") + 
#   theme_bw()
# 
# labeled_sptree <- sptree
# labeled_sptree$tip.label <- tax$Species[match(labeled_sptree$tip.label, tax$mnemo)]
# 
# sptree_plot <- ggtree(labeled_sptree) + 
#   geom_tiplab(aes(label=sub("[^A-Z].+? ", ". ", label)),fontface="italic",
#               align = TRUE) + 
#   geom_nodelab(color="firebrick", 
#                nudge_x = -.15, hjust = 1, 
#                nudge_y = .3, size=3.5)+
#   xlim(c(-4, 21))
# 
# fig_s2 <- (sptree_plot + pp1_plot) +
#   plot_layout(widths = c(1, 1.5)) +
#   plot_annotation(tag_levels = 'A') &
#   theme(plot.tag = element_text(size = 10, face = "bold", family = "Helvetica"))
# ggsave("draft/figures/figS2.pdf", fig_s2, width = 12, height = 6, device = cairo_pdf)
<!-- ``` -->

# Review plots

### Experimental vs predicted

```{r why_no_exp}

plot_PDB <- threed_df %>%
  group_by(mnemo) %>% 
  filter(!is.na(mnemo)) %>% 
  summarise_all(~ (1-sum(is.na(.))/n())*100) %>% 
  left_join(select(tax, mnemo, Species)) %>% 
  ggplot(aes(PDB, fct_reorder(Species, PDB))) + 
  geom_text(aes(label=signif(PDB,2)), family="Aptos", 
            size=2.5, nudge_x = 1, hjust = 0) + 
  geom_bar(stat="identity") + 
  scale_x_continuous(limits = c(0,100), expand = expansion(add = c(0,2))) +
  labs(y="Species", x="% Proteins with PDB entry") + 
  theme(axis.text.y = element_text(face = "italic"))

 
ggsave("draft/figures/review_nopdb.png", plot_PDB, width = 4, height = 3, dpi = 600)
```

### Paralogy issue in phylome

```{r}
# find results/phylogeny/hsap_1kseeds/seeds/UP000005640/ -name "*.ids" -print0 | while IFS= read -r -d '' file; do  awk -v fname="$(basename "$file" | sed 's/.ids//' | sed 's/_/\t/')"  '{print $0 "\t" fname}' "$file"; done > test/paranoia/ids_by_set.txt
df_ids <- read_delim("draft/figures/ids_by_set.txt", 
                 col_names = c("target", "query", "set")) %>% 
  filter(query %in% aln_seeds)

overlap_df <- filter(df_ids, query!=target) %>% 
  group_by(query, set) %>% 
  summarise(targets=list(target)) %>% 
  # head(100) %>%
  cross_join(., .) %>% 
  filter(set.x==set.y, query.x!=query.y) %>% 
  rowwise() %>% 
  mutate(n =  length(union(targets.x, targets.y)), 
         overlap = length(intersect(targets.x, targets.y))/n) %>% 
  select(query.x, query.y, overlap, n, set.x) %>% 
  filter(query.x>query.y)

overlap_df %>% 
  mutate(is_not0 = overlap>0) %>% 
  count(set.x, is_not0) %>% 
  pivot_wider(names_from = is_not0, values_from = n) %>% 
  mutate(proportion = `TRUE`/(`FALSE`+`TRUE`)*100)

```

### Heteroscedascity of DL normalization

```{r}
lm <- lm(dups+losses ~ n_tips, data = reco)
reco$res <- lm$residuals

tmp_plot <- ggplot(reco, aes(y = res, x = n_tips, fill=target)) + 
  geom_point(pch=21) +
  facet_wrap(~target) + 
  geom_abline(slope = 0) + 
  scale_fill_manual(values = palettes_method) + 
  labs(x="Num. of Tips", 
       y="Residuals of: D+L ~ Num. of Tips") + 
  theme(legend.position = "none")

ggsave("draft/figures/review_hetero_TMP.png", tmp_plot,
       width = 4, height = 3)
```

