---
title: "LUCA sampling"
# knit: (function(input_file, encoding) {
#   rmarkdown::render(input_file,
date: "`r format(Sys.time(), '%d-%b-%Y')`"
author: "Giacomo Mutti"
output:
    rmdformats::html_clean:
      number_sections: false
      code_folding: hide
      self_contained: true
      thumbnails: false
      lightbox: false
    highlight: kate
---

<style>
h1, h2, h3, h4, h5, h6, legend {
    font-family: Helvetica, sans-serif;
    font-weight: 700;
    color: #00a276;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_hooks$set(crop = knitr::hook_pdfcrop)
knitr::opts_chunk$set(crop = TRUE)
```

```{r libraries, message=FALSE, warning=FALSE}
library(ggtree)
library(ape)
library(tidyverse)
library(patchwork)
library(kableExtra)
source("scripts/functions.R")
```

# Prokaryotes

In this notebook we will get a representative prokaryotic taxon sampling for a LUCA dataset to compare deep homologies inferred by blast or foldseek. In order to do this we can use **[GTDB](https://gtdb.ecogenomic.org/)**. It has selected reference genomes for both *Bacteria* and *Archaea*. There is a species tree available for both kingdoms: 

* **[bacterial tree](https://data.gtdb.ecogenomic.org/releases/latest/bac120.tree)**
* **[archaeal tree](https://data.gtdb.ecogenomic.org/releases/latest/ar53.tree)**

Each tree with associated metadata:

* **[bacterial metadata](https://data.gtdb.ecogenomic.org/releases/latest/bac120_metadata.tsv.gz)**
* **[archaeal metadata](https://data.gtdb.ecogenomic.org/releases/latest/ar53_metadata.tsv.gz)**

As all genomes in uniprot have structural predictions we also need to check if the genomes are available. We can look in the **[uniprot genomes](https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/STATS)** table.


```{r params, class.source = 'fold-show'}
min_completeness <- 90
max_contamination <- 5
stepwise_species <- 1/10
target_species <- 20
```
```{r files}
# meta
uniprot_genomes <- "data/meta/uniprot_genomes.tsv"
uniprot_ref <- "data/meta/uniprot_ref.tsv"
ar_meta_file <- "data/meta/ar53_metadata.tsv.gz"
bac_meta_file <- "data/meta/bac120_metadata.tsv.gz"
taxids_in_AF <- "data/meta/taxids_in_alphafold.txt"
# LUCA eukaryotes
euka <- "data/input_tables/Hsapopi.csv"
# trees
bac120_file <- "data/sptrees/bac120.tree"
ar53_file <- "data/sptrees/ar53.tree"
```

```{r read_files, cache=TRUE}
taxids_AF <- as.numeric(readLines(taxids_in_AF))

up_ref <- read_delim(uniprot_ref, show_col_types = FALSE)
up_genomes_raw <- read_delim(uniprot_genomes, show_col_types = FALSE, col_types = c("cncccccc")) %>% 
  left_join(select(up_ref, Proteome_ID, SUPERREGNUM), by = "Proteome_ID")

up_genomes <- up_genomes_raw %>%
  mutate(assembly=gsub(" .*", "", `Assembly(Previous_Assembly)`)) %>%
  filter(Proteome_ID %in% up_ref$Proteome_ID, Tax_ID %in% taxids_AF) %>% 
  select(Proteome_ID, Species_Name, assembly) 

bac <- read.tree(bac120_file)
ar <- read.tree(ar53_file)

gtdb_cols <- c("accession", "checkm_contamination", "checkm_completeness",
               "genome_size", "gtdb_taxonomy", "gtdb_representative",
               "ncbi_assembly_level", "ncbi_genbank_assembly_accession",
               "ncbi_organism_name", "protein_count")

bac_meta <- read_delim(bac_meta_file, show_col_types = FALSE) %>%
  select(all_of(gtdb_cols)) %>%
  filter(gtdb_representative,
         checkm_completeness >= min_completeness,
         checkm_contamination <= max_contamination,
         ncbi_genbank_assembly_accession %in% up_genomes$assembly)

asgards <- c("GCA_016933575.1", "GCA_008000775.1")

ar_meta <- read_delim(ar_meta_file, show_col_types = FALSE) %>%
  select(all_of(gtdb_cols)) %>%
  filter((gtdb_representative & checkm_completeness >= min_completeness & checkm_contamination <= max_contamination &
         ncbi_genbank_assembly_accession %in% up_genomes$assembly) | ncbi_genbank_assembly_accession %in% asgards)

```

There are **`r length(bac$tip.label)`** reference bacterias and **`r length(ar$tip.label)`** archaeas in the species trees.

Of these, there are **`r length(unique(bac_meta$ncbi_genbank_assembly_accession))`** Bacterias and **`r length(unique(ar_meta$ncbi_genbank_assembly_accession))`** Archaea that are in **Uniprot** and **AF** and respect the completeness (greater then **`r min_completeness`** and contamination (lower than **`r max_contamination`**) criteria. Notably, as they are crucial, we included the two Asgard genomes shared by GTDB and Uniprot: `r paste(asgards , collapse=",")`, even though they did not pass the checkM criterias (although not by much).

We therefore subset the species trees with these genomes. We can use **[Treemer](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2164-8)**: a tree pruning algorithm optimized to keep as much as information as possible. the `-r` parameter is what determines the speed of the process. We decided to use 1/250 of the leaves (if 1000 leaves, remove 4 tips before recomputing distances, which is the slow part). We will keep `r target_species` tips. With `-mc -lm` option we can force treemer to include all asgards genomes.

```{r intermediate_trees}
ar_intermediate <- drop.tip(ar, ar$tip.label[-match(ar_meta$accession, ar$tip.label)])
bac_intermediate <- drop.tip(bac, bac$tip.label[-match(bac_meta$accession, bac$tip.label)])

mc <- length(asgards)
asgard_tips <- ar_meta[ar_meta$ncbi_genbank_assembly_accession %in% asgards, ]$accession
write_csv(tibble(asgard_tips, rep("Asgard", mc)), paste0(ar53_file, ".filtered.lm"), col_names = FALSE)

write.tree(ar_intermediate, paste0(ar53_file, ".filtered"))
write.tree(bac_intermediate, paste0(bac120_file, ".filtered"))
```

```{r treemer, cache=TRUE}
py="python3"
ar_resolution <- round(length(ar_intermediate$tip.label)*stepwise_species)
args=paste("scripts/Treemmer_v0.3.py", paste0(ar53_file, ".filtered"),
           "-X",target_species,"-r",ar_resolution,"-lm",paste0(ar53_file, ".filtered.lm"),"-mc",mc,"-np -v 0 -lp 0")
system2(command=py, args=args)
bac_resolution <- round(length(bac_intermediate$tip.label)*stepwise_species)
args=paste("scripts/Treemmer_v0.3.py", paste0(bac120_file, ".filtered"),
           "-X",target_species,"-r",bac_resolution,"-np -v 0 -lp 0")
system2(command=py, args=args)
```

### Reduced trees

```{r plot trees, fig.align = "center", fig.width=14, fig.height=8, out.width="100%"}
ar_final_file <- paste0(ar53_file, ".filtered_trimmed_tree_X_", target_species)
ar_final <- read.tree(ar_final_file)

bac_final_file <- paste0(bac120_file, ".filtered_trimmed_tree_X_", target_species)
bac_final <- read.tree(bac_final_file)

ar_plot <- fortify(ar_intermediate) %>%
  left_join(ar_meta, by = c("label"="accession")) %>%
  mutate(is_selected=label %in% ar_final$tip.label, is_asgard=label %in% asgard_tips) %>% 
  ggtree(layout = "circular") +
  ggtitle("Archaea selected genomes") +
  geom_tippoint(aes(color=checkm_completeness, shape=is_asgard), . %>% filter(is_selected), size=4) +
  xlim(c(0, max(vcv(ar_intermediate)))) +
  scale_color_gradientn(colours = wespal, limits=c(min_completeness-5, 100))

bac_plot <- fortify(bac_intermediate) %>%
  left_join(bac_meta, by = c("label"="accession")) %>%
  mutate(is_selected=label %in% bac_final$tip.label) %>% 
  ggtree(layout = "circular") +
  ggtitle("Bacteria selected genomes") +
  geom_tippoint(aes(color=checkm_completeness), . %>% filter(is_selected), size=4) +
  xlim(c(0, max(vcv(bac_intermediate)))) +
  scale_color_gradientn(colours = wespal, limits=c(min_completeness-5, 100))

trees_plot <- ar_plot + bac_plot & theme(legend.position = "bottom")
trees_plot + plot_layout(guides = "collect")
```

```{r final_table}
ar_final_table <- filter(ar_meta, accession %in% ar_final$tip.label) %>% 
  left_join(up_genomes, by=c("ncbi_genbank_assembly_accession"="assembly")) %>% 
  mutate(phylum=gsub(";.*", "", gsub(".*;p__", "", gtdb_taxonomy)), kingdom="Archaea") %>%
  select(Proteome_ID, ncbi_genbank_assembly_accession, Species_Name, phylum,
         protein_count, genome_size, checkm_completeness, checkm_contamination) %>% 
  arrange(phylum)

bac_final_table <- filter(bac_meta, accession %in% bac_final$tip.label) %>% 
  left_join(up_genomes, by=c("ncbi_genbank_assembly_accession"="assembly")) %>% 
  mutate(phylum=gsub(";.*", "", gsub(".*;p__", "", gtdb_taxonomy)), kingdom="Bacteria") %>% 
  select(Proteome_ID, ncbi_genbank_assembly_accession, Species_Name, phylum,
         protein_count, genome_size, checkm_completeness, checkm_contamination) %>% 
  arrange(phylum)

rbind(ar_final_table, bac_final_table) %>% 
  kbl(booktabs = TRUE, longtable = TRUE) %>% 
  kable_styling(bootstrap_options = "striped",
              font_size = 9,
              latex_options = c("hold_position","repeat_header"),position = "left")
```

# Run LUCA analysis

We will add then these 40 proteomes to the 18 Eukaryotes of the *H. sapiens Opisthokonta* dataset and create the file *data/input_tables/LUCA.csv* that will be then used in the pipeline.

```{r create_input_table}
read_delim(euka, delim = "\t", col_names = colnames(up_genomes_raw), show_col_types = FALSE) %>% 
  rbind(up_genomes_raw[up_genomes_raw$Proteome_ID %in% ar_final_table$Proteome_ID, ]) %>% 
  rbind(up_genomes_raw[up_genomes_raw$Proteome_ID %in% bac_final_table$Proteome_ID, ]) %>% 
  write_delim("data/input_tables/LUCA.csv", delim = "\t", col_names = FALSE)
```

First you need to download the structures, do this in local:

```{bash download, eval=FALSE, class.source = "fold-show"}
snakemake --configfile data/configs/LUCA.yaml -p -j2 --until download_pdbs -s workflow/download_data.smk
snakemake --configfile data/configs/LUCA.yaml -p -j2 --until download_uniprot_ids -s workflow/download_data.smk
```

Then run on MN the remaining steps to decompress:

```{bash untar, eval=FALSE, class.source = "fold-show"}
snakemake --configfile data/configs/LUCA.yaml -p -j24 -s workflow/download_data.smk
```

Finally, run the pipeline until the homology call:

```{bash smk, eval=FALSE, class.source = "fold-show"}
snakemake --configfile data/configs/LUCA.yaml -p -j24 --until plot_evalues
```

